pkg/mcp/internal/test/rate_limiter.go

```diff
@@ -0,0 +1,41 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package test
+
+import (
+	"context"
+)
+
+type FakeRateLimiter struct {
+	WaitErr chan error
+	WaitCh  chan context.Context
+}
+
+func NewFakeRateLimiter() *FakeRateLimiter {
+	return &FakeRateLimiter{
+		WaitErr: make(chan error),
+		WaitCh:  make(chan context.Context, 10),
+	}
+}
+
+func (f *FakeRateLimiter) Wait(ctx context.Context) error {
+	f.WaitCh <- ctx
+	select {
+	case err := <-f.WaitErr:
+		return err
+	default:
+		return nil
+	}
```

Comment:
Won't this race if the stream invokes `Wait()` before the test code can write an error to `WaitErr`?

Reply:
There could be a potential for race when expecting an error but not providing the error prior to stream invoking `Wait()`, however since this a fake the caller is in a full control of the behaviour, I don't see a big concern.

Reply:
I was looking at the usage [below](https://github.com/istio/istio/pull/11435/files#diff-ff91956ecb2abc8f1ced2ce40ddf6434R71) where the stream is invoked via a goroutine and then an error is written to the WaitErr channel. It seems possible for the goroutine to run first and hit line 39 before the WaitErr is written.